notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+       "datetime.datetime(2023, 10, 26, 0, 59, 16, 892032)"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# ingest users\n",
+    "sql='''\n",
+    "REPLACE INTO \"users\" OVERWRITE ALL\n",
+    "WITH \"ext\" AS (SELECT *\n",
+    "FROM TABLE(\n",
+    "  EXTERN(\n",
+    "    '{\"type\":\"http\",\"uris\":[\"http://datagen:9999/file/users.json\"]}',\n",
+    "    '{\"type\":\"json\"}'\n",
+    "  )\n",
+    ") EXTEND (\"time\" VARCHAR, \"user_id\" VARCHAR, \"first_name\" VARCHAR, \"last_name\" VARCHAR, \"dob\" VARCHAR, \"address_lat\" VARCHAR, \"address_long\" VARCHAR, \"marital_status\" VARCHAR, \"income\" VARCHAR, \"signup_ts\" VARCHAR))\n",
+    "SELECT\n",
+    "  TIMESTAMP'1970-01-01 00:00:00' AS \"__time\",\n",
+    "  \"user_id\",\n",
+    "  \"first_name\",\n",
+    "  \"last_name\",\n",
+    "  \"dob\",\n",
+    "  \"address_lat\",\n",
+    "  \"address_long\",\n",
+    "  \"marital_status\",\n",
+    "  \"income\",\n",
+    "  \"signup_ts\"\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY ALL\n",
+    "'''\n",
+    "display.run_task(sql)\n",
+    "print_timestamp()\n",
+    "\n",
+    "# ingest clicks\n",
+    "sql='''\n",
+    "REPLACE INTO \"clicks\" OVERWRITE ALL\n",
+    "WITH \"ext\" AS (SELECT *\n",
+    "FROM TABLE(\n",
+    "  EXTERN(\n",
+    "    '{\"type\":\"http\",\"uris\":[\"http://datagen:9999/file/clicks.json\"]}',\n",
+    "    '{\"type\":\"json\"}'\n",
+    "  )\n",
+    ") EXTEND (\"time\" VARCHAR, \"user_id\" VARCHAR, \"event_type\" VARCHAR, \"client_ip\" VARCHAR, \"client_device\" VARCHAR, \"client_lang\" VARCHAR, \"client_country\" VARCHAR, \"referrer\" VARCHAR, \"keyword\" VARCHAR, \"product\" VARCHAR))\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"time\") AS \"__time\",\n",
+    "  \"user_id\",\n",
+    "  \"event_type\",\n",
+    "  \"client_ip\",\n",
+    "  \"client_device\",\n",
+    "  \"client_lang\",\n",
+    "  \"client_country\",\n",
+    "  \"referrer\",\n",
+    "  \"keyword\",\n",
+    "  \"product\"\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "display.run_task(sql)\n",
+    "print_timestamp()\n",
+    "\n",
+    "# make sure both tables are available before moving on\n",
+    "sql_client.wait_until_ready('users')\n",
+    "sql_client.wait_until_ready('clicks')\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "472589e4-1026-4b3b-bb79-eedabb2b44c4",
+   "metadata": {},
+   "source": [
+    "### Pre-join into Denormalized Table\n",
+    "In many cases it is best to pre-join the data. Druid is particularly suited for analytic applications that require subsecond response times in order to drive an interactive user experience with high concurrency. In a parallel system, avoiding the need to move the data that comes with a join at query time, is one of the tools that is useful to achieve high query performance at high concurrencies."
```

Comment:
As above, suggest you rephrase this out of advisory and coaching land.

Check out this SQL. It makes a salad. It does this by:

* Tossing the leaves
* Adding beans and sausages together.

Reply:
That's exactly what this sequence does.... it loads them separately and then loads them into a single table, pre-joining them.
I think the comments here are important and not meant so much to coach, but to understand what Druid is designed to do and that pre-joining the data is not a bad thing. 

The point of the sequence is that we first prepare all the data in the different forms,  separate tables, pre-joined, in a lookup, and then shows how to process the same queries using joins, lookups and pre-join.  I don't think that changing the order or adding more SELECTs in between helps the narrative.... convince me otherwise ;-)